Add built-in procedure for changing password

henriknyman opened the pull request:
This PR changes how Bolt handles when the authenticated user is required to change password. 
Instead of closing the session, it is kept alive to allow calling a new built-in procedure sys.auth.changePassword. Any transactions opened will have access mode NONE, which only allows access to a new type of dbms operations (currently the only operation is procedureCallDbms).

We now tie the authenticated subject that implements the AccessMode interface to a transaction, instead of having just a static AccessMode.

henriknyman committed:
Change AccessMode to an interface

Commit:
Move AccessMode to common

Commit:
Added inbuilt proc allowing a user to change their password

henriknyman committed:
Fix handling password change required on login

henriknyman committed:
Fix issues with change password procedure

- Add Bolt integration tests for the new built-in change password procedure
- Fix access mode for beginImplicitTransaction in Bolt session state machine
- Expose restrict access mode on Neo4jTransactionContext
- Change access mode to READ during query planning
- Make tokenNameLookup a lazy val in TransactionBoundQueryContext
to be able to execute queries without having read access

henriknyman committed:
Fix bolt BasicAuthentication tests

boggle commented on community/kernel/src/main/java/org/neo4j/kernel/builtinprocs/BuiltInProcedures.java:
```diff
@@ -52,5 +52,8 @@ public void accept( Procedures procs ) throws ProcedureException
         procs.register( new ListComponentsProcedure( procedureName( "sys", "components" ), neo4jVersion ) );
+
+        // These are 'sys.auth'-namespaced procedures; these deal with authentication and authorization-oriented operations
+        procs.register( new AlterUserPasswordProcedure( procedureName( "sys", "auth", "changePassword" ) ) );
```
We talked this over in the drivers and we'd like to not introduce nested namespaces right now, please change to `sys.changePassword`

We can pick this discussion up again once we add more `auth` related functions.

henriknyman replied:
Done

technige commented on community/kernel/src/test/java/org/neo4j/kernel/builtinprocs/BuiltInProceduresTest.java:
```diff
@@ -143,6 +145,7 @@ public void shouldListCorrectBuiltinProcedures() throws Throwable
             record( "db.labels", "db.labels() :: (label :: STRING?)" ),
             record( "db.propertyKeys", "db.propertyKeys() :: (propertyKey :: STRING?)" ),
             record( "db.relationshipTypes", "db.relationshipTypes() :: (relationshipType :: STRING?)" ),
+            record( "sys.auth.changePassword", "sys.auth.changePassword(password :: STRING?) :: ()" ),
```
Too early to sub-namespace these. Let's stick with `sys.changePassword` for now and add more levels when we have significantly more built-ins. The presence of the word "password" already suggests this is auth-related.

henriknyman replied:
Done

technige commented on community/common/src/main/java/org/neo4j/kernel/api/security/AccessMode.java:
```diff
@@ -0,0 +1,142 @@
+    enum Static implements AccessMode
+    {
+        /** No reading or writing allowed. */
+        NONE
```
Strictly speaking, is this `NONE` or `PASSWORD_CHANGE_ONLY`? If the latter, we should name it as such. If the former, what is the value of being allowed in with absolutely no access, as opposed to not being allowed in?

henriknyman replied:
A user being allowed to change his own password is implemented as a more fine-grained permission on the AuthSubject (which is now something like the "security context" tied to the transaction, even though the kernel statement operations currently only views it through the AccessMode interface)

henriknyman committed:
Change the Bolt result for credentials expired

- On password change required the Bolt server now responds with
<SUCCESS { "credentials_expired": true}>
- Changed AuthSubject.setPassword to void and throwing exceptions on failure
- Changed the changePassword built-in procedure to return an empty result

henriknyman committed:
Remove nested namespace for built-in procedure

 - Renames sys.auth.changePassword -> sys.changePassword

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/builtinprocs/AlterUserPasswordProcedure.java:
```diff
@@ -0,0 +1,75 @@
+
+        try
+        {
+            authSubject.setPassword( input[0].toString() );
```
should we add some restrictions on what password should be?

henriknyman replied:
I think this should be up to the AuthManager implementation (and to throw an appropriate exception). Unfortunately the current legacy BasicAuthManager have no restrictions, and the few restrictions we have are instead implemented directly in the REST UserService.
For enterprise we will probably have requirements for password policies so I think we can deal with it later in the new auth manager.

henriknyman committed:
Remove DbmsOperations from KernelStatement

- QueryTransactionalContext is extended with dbms operations
- DbmsOperations are made independent of the kernel transaction as a separate
dependency.
- A NonTransactionalDbmsOperations implementation is provided by
DataSourceModule.
- The context for dbms procedure calls are set up with only AuthSubject
(as AccessMode), and no kernel transaction. (This context could instead
be extended in the future with a new type of dbms transaction when needed.)

henriknyman committed:
Move dbms operations into dbms sub-package
